Shapes: Add simple border option

Trapsta opened the pull request:
## Context

Enables the border panel for non-rectangular shapes with the possibility to add a simple border that supports:

- A single border-width value only.
- Allows choosing color **but not** opacity for the border.
- Only a middle-positioned border (unlike for rectangular elements).

This is all tested and confirmed working with output, flip, rotation, focal changes, focal zoom, page zoom, RTL, layer opacity, and more.

### Todo

- [x] Fix media snapping bug (currently media has a hard time snapping into shapes - not sure what broke) Must be fixed in this PR!
- [x] Fix media edit mode when in non-rect bordered shape
- [x] Fix the weird scaling bug
- [ ] Figure out what happens when dropping a multi-bordered and/or semitransparent-bordered media element into a shape that does not support it (follow-up bug probably).
- [ ] File follow-up bug for the color picker opacity error text discrepancy mentioned below.

## User-facing changes
### Design panel
The new border width panel for non-rectangular shapes (note the single input, no lock button and no opacity for border color):
![Screen Shot 2022-01-22 at 00 02 47](https://user-images.githubusercontent.com/637548/150625464-f4ec8138-e110-4a5c-9b0d-a639133c85bd.png)

Note the following "bug" in the basic color picker, where transparency is not allowed (which is correct), but the error message assumes that this is because it's about setting page background color, which this isn't.
![Screen Shot 2022-01-22 at 00 02 57](https://user-images.githubusercontent.com/637548/150625517-f278000b-1f1c-4be6-83d7-4b3356d92bb4.png)

A follow-up bug should be filed when this is merged and the text should be customized for page background and non-rectangular shape/mask border-color respectively. Or it should be made generic enough to support both cases (e.g. _"The color on this element does not support opacity"_).

### Canvas
Here are the various elements with their borders increasing and decreasing again:

| ![border-rect](https://user-images.githubusercontent.com/637548/150626183-87f4cf10-bd71-4192-8979-dad461d04dbe.gif) | ![border-circ](https://user-images.githubusercontent.com/637548/150626187-05d379a9-23e3-4c69-9e44-910e5d1d8e0f.gif) | ![border-tri](https://user-images.githubusercontent.com/637548/150626190-6cc77091-6257-4fd3-b613-79da088bb215.gif) | ![border-heart](https://user-images.githubusercontent.com/637548/150626191-40b9f62e-9d0b-4808-9e04-963c68172b33.gif) |
|-|-|-|-|
| ![border-star](https://user-images.githubusercontent.com/637548/150626238-3a4b1a49-f0ff-4bb7-8e36-c263d496988b.gif) | ![border-pent](https://user-images.githubusercontent.com/637548/150626239-7e625d99-567f-4998-b721-66bcbb125e08.gif) | ![border-hex](https://user-images.githubusercontent.com/637548/150626240-456aba8e-f7f0-48be-afa3-9b6a6578f47f.gif) | ![border-blob1](https://user-images.githubusercontent.com/637548/150626241-17c91764-b6b6-4422-bcd0-814d8b8503b6.gif) |
| ![border-blob2](https://user-images.githubusercontent.com/637548/150626244-06d97e40-7714-4b9b-84cc-debd9f23f90a.gif) | ![border-blob3](https://user-images.githubusercontent.com/637548/150626246-1267537d-eef6-4e69-aaeb-21278467d0ca.gif) |![border-blob4](https://user-images.githubusercontent.com/637548/150626247-1b2f29f3-00aa-47fd-a5f7-6dbbb2c6eabb.gif) |![border-blob5](https://user-images.githubusercontent.com/637548/150626249-63ee538d-2b31-4474-a0fd-f6a66aecf265.gif) |

### Output
Compare canvas with output:
| Canvas | Output |
|-|-|
| ![Screen Shot 2022-01-22 at 00 08 49](https://user-images.githubusercontent.com/637548/150625579-99638576-4e40-4d21-adb2-6bacba3c78a2.png) | ![Screen Shot 2022-01-22 at 00 08 59](https://user-images.githubusercontent.com/637548/150625580-cfa249e6-8fba-49df-8b61-974626b8908f.png) |

## Testing Instructions

This PR can be tested by following these steps:

1. Add  a non-rectangular shape to the page
2. Select the shape and verify that you can edit the border width  from the border panel
3. verify that you can update the border width and color, but not the opacity
4. observe that you cannot toggle the border width to input multiple values
5. set layer opacity to 50% and observe that the whole element correctly fades down and there's no weirdness happening in the border overlap region

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6171

Trapsta committed:
chore(web-stories-wp): update non-rectanguar shapes supportsborder property

Closes #web-6171

Trapsta committed:
refactor(web-stories-wp): disable border width consistent values toggle for non-rectangular shapes

Closes #web-6171

Trapsta committed:
fix(web-stories-wp): skip getborderstyle for non-rectangular shapes

Closes #web-6171

Trapsta committed:
feat(web-stories-wp): add clipped svg border for non-rect shapes

Closes #web-6171

Trapsta committed:
fix(web-stories-wp): fix for non-rect shapes border in display , page thumbnails & preview

Closes #web-6171

barklund commented on packages/story-editor/src/components/canvas/displayElement.js:
```diff
@@ -70,6 +70,10 @@ const Wrapper = styled.div.attrs(
+
+      border-radius: ${theme.borders.radius.small};
+      overflow: hidden;
+    `}
```
This looks like the same annoying linter "fixes" that I've also experienced. Try removing these lines locally (only this) and commit it. If the linter adds it again and then says "nothing to commit", it's clearly the same issue.

A short-term workaround is to remove it in a commit and do "git commit --no-verify" to circumvent the linters. But as soon as you edit this file again, the linter will re-add this crap in pre-commit hook.

Long-term we really need to figure out why this happens.

swissspidy replied:
@barklund Can you please create a ticket for this? Thanks.

barklund replied:
* Done in https://github.com/google/web-stories-wp/issues/9865

barklund commented on packages/story-editor/src/components/panels/design/border/borderWidth.js:
```diff
@@ -71,8 +72,13 @@ function WidthControls({ selectedElements, pushUpdateForObject }) {
+  // Some shapes i.e all non-rectangular shapes only support a single border width value
+  const singleBorderSupport = selectedElements.some((el) =>
+    singleBorderMask(el)
+  );
```
This can be collapsed to just this:

```js
  const singleBorderSupport = selectedElements.some(singleBorderMask);
```

I would also prefer the function (and the variable) having a "has" or "is" prefix given that it returns (or is) a boolean, just to make it clear. So maybe just "hasSingleBorderedMask" or something like that? We might have a similar function elsewhere that checks for non-rectangular masks.

barklund replied:
Maybe the two are even misnamed? All shapes support a single border, the issue is that non-rectangular shapes don't support multiple border values, so perhaps it should be:

 ```js
  const hasMultiBorderSupport = selectedElements.every(hasMultiBorderedMask);
```

And then reverse the logic throughout. I feel that makes more sense when skimming the code later.

Trapsta replied:
Yeah, I agree with that, will make the variables names more meaningful and easy to understand. On it

barklund commented on packages/story-editor/src/utils/elementBorder.js:
```diff
@@ -194,3 +198,28 @@ export function getResponsiveBorder(border, previewMode, converter) {
+export function getBorderWrapperStyle(maskId, borderWidth) {
+  if (!maskId || !borderWidth) {
+    return {};
+  }
```
Could this check for the rectangular mask as well and just return nothing here?

Trapsta replied:
Yes, it could

barklund reviewed (commented):
Just some minor nits on style, but also a linter issue.

miina commented on packages/story-editor/src/masks/output.js:
```diff
@@ -77,35 +89,66 @@ export default function WithMask({
 
+  const borderWidth = border?.left;
+  const showSingleBorder = Boolean(singleBorderMask(element) && borderWidth);
+  const borderColor = border?.color
```
If there's no `borderColor`, should we still display the border at all?

Agreed with Morten that `singleBorderMask` isn't the most self-explanatory name here, and it would be better to rename it to something like `hasMultiBorderSupport` or `canHaveMultiBorder` or sth similar.

Probably just using `showBorder` is ok here as well for the name.

Trapsta replied:
Okay, gotcha

miina commented on packages/story-editor/src/masks/display.js:
```diff
@@ -85,28 +107,53 @@ export default function WithMask({
-            clipPathUnits="objectBoundingBox"
+    <>
+      {showSingleBorder && (
+        <div style={{ ...style, ...wrapperStyle }}>
```
Is there difference between the `display` and `output` for displaying the SVG border? If not, then maybe we could even move this to an util as well, e.g. `getMaskBorder` or something like that. WDYT?

Trapsta replied:
yeah, no difference really so moving to a util would eliminate some duplication

Trapsta committed:
Merge branch 'main' into feat/web-stories-6171-simple-border-for-non-rectangular-shapes

Trapsta committed:
correct linting error

Trapsta committed:
Merge branch 'feat/web-stories-6171-simple-border-for-non-rectangular-shapes' of github.com:google/web-stories-wp into feat/web-stories-6171-simple-border-for-non-rectangular-shapes

Commit:
Merge branch 'main' into feat/web-stories-6171-simple-border-for-non-rectangular-shapes

# Conflicts:
#	packages/story-editor/src/components/panels/design/border/borderWidth.js

Commit:
Failed attempt at using SVG border with punchout

Commit:
Merge branch 'main' into feat/web-stories-6171-simple-border-for-non-rectangular-shapes

Commit:
Fixed overlay borders on non-rectangular shapes

Commit:
Fixed output and changed to relative CSS instead

Commit:
Fixed flipped transforms too

Commit:
Fixed flip in the output too

Commit:
Fixed image edit mode

miina commented on packages/story-editor/src/masks/output.js:
```diff
@@ -77,16 +94,86 @@ export default function WithMask({
 
   const maskId = `mask-${mask.type}-${element.id}-output`;
 
+  if (showSingleBorder) {
```
Same question as before, should we show this at all if `borderColor === 'none'`?

Maybe something like this:
```
    const borderWidth = border?.left;
    const borderColor = border?.color ? getBorderColor({ color: border.color }) : undefined;
  const showSingleBorder = Boolean(
    !checkMultiBorderSupport(element) && borderWidth && borderColor
  );
```

barklund replied:
Simplified it in ce4980aea.

miina commented on packages/story-editor/src/masks/index.js:
```diff
@@ -43,3 +48,34 @@ function getDefaultElementMask(type) {
 }
+
+const BORDER_MULTIPLIER = 3;
+export function getBorderedMaskProperties(
```
Would be great to leave a general comment somewhere explaining how this works.

There's a slight difference currently with "normal" borders -- the frame line is shown in the middle of the border as opposed to around the origin element. Probably this could be unified in #5325 where all the border will be included inside the frame for all elements?
<img width="243" alt="Screenshot 2022-01-24 at 11 57 32" src="https://user-images.githubusercontent.com/3294597/150761812-c5b79269-d5e6-4f0f-9349-3d0b54462371.png">

barklund replied:
Oh, the middle-aligned border (rather than the outer border) is a significant change, but it's the only way this could work. It changes workflow quite a bit for end-users and is definitely undesirable.

But there is no other way to do this, as it is way too hard to synchronize the two SVG's on top of each other if we want to "punch out" the shape of the element "inside" the border. It just didn't work reliably enough to be useful.

So this is what we have. It's this or nothing, and I feel that this is better.

miina replied:
Ah, yes, definitely, just wanted to confirm mainly if this will still work for #5325 as well (where the frame would be moved to outside line of the border) -- to make the border clickable, too.

barklund replied:
Added documentation in 4193c5d1b.

miina commented on packages/story-editor/src/masks/index.js:
```diff
@@ -32,6 +32,11 @@ export function canMaskHaveBorder(element) {
   return !mask || mask.supportsBorder;
 }
 
+export function checkMultiBorderSupport(element) {
```
Nit: Would be nice to have this start with `can...` as well, e.g. `canHaveMultiBorder`. `check...` doesn't really say it will return a boolean, starting with `has...` or `can...` would be great.

barklund replied:
Renamed to `canSupportMultiBorder()` in 9165f3f9e.

miina commented on packages/story-editor/src/masks/display.js:
```diff
@@ -85,6 +114,68 @@ export default function WithMask({
     previewMode ? '-preview' : ''
   }-${randomId}`;
 
+  if (showSingleBorder) {
```
If this is exactly the same as in `output.js`,  could be move it to a helper (DRY)?

barklund replied:
This was a tough refactor (so many things were similar but not exactly the same), but managed it in 1c2a42a8c

miina reviewed (commented):
🎉 Omg, this seems to work! 🎉 🌊 

Some unit tests are failing now and would be good to have Karma added to the current border Karma, too.

There's actually one case that is not working currently (due to the border partially covering the element itself as well) -- when the element is small and border is huuuuge:
<img width="280" alt="Screenshot 2022-01-24 at 12 07 19" src="https://user-images.githubusercontent.com/3294597/150763031-8e6478b7-4114-4a4b-bc6a-492af14d7692.png">

This is quite an edge case so maybe something to confirm with UX -- e.g. if we should just leave it / limit the max border width, thoughts?

miina reviewed (approved):
Approving (pending the open comments) based on the discussion with product/UX it's ok to have the edge cases and follow up separately (not spending too much extra time though).

Commit:
Converted actual border to non-scaling-stroke reversing the calculation

Commit:
Merge branch 'main' into feat/web-stories-6171-simple-border-for-non-rectangular-shapes

Commit:
Added documentation for the border mask expansion logic

Commit:
Refactored mask logic to single component

Commit:
Renamed utility function

Commit:
Simplified logic for when to show fancy border

Commit:
Fixed karma test for non-rect shape border

Commit:
Fixed some potential bugs in bordered element

Commit:
Fixed unit tests now that non-rect shapes can have border (but not radius)

Commit:
Removed border radius from non-rect shapes

The input was shown, but did not actually do anything as non-rectangular shapes just have a border that follow the shape regardless of the radius setting.

Commit:
Stopped passing unused prop to html element

Commit:
Only included transform if non-empty

Commit:
Fixed test for border radius output on non-rect shape
